app/jobs/course/assessment/answer/base_auto_grading_job.rb

```diff
@@ -1,31 +1,61 @@
 # frozen_string_literal: true
 class Course::Assessment::Answer::BaseAutoGradingJob < ApplicationJob
-  private
+  include TrackableJob
 
-  def downgrade_if_overdue(question, execution_time)
-    default_timeout = Course::Assessment::ProgrammingEvaluationService::DEFAULT_TIMEOUT
-    return unless !question.is_low_priority? && execution_time > default_timeout
+  DEFAULT_TIMEOUT = Course::Assessment::ProgrammingEvaluationService::DEFAULT_TIMEOUT
 
-    question.modify_is_low_priority!
+  retry_on PriorityShouldBeLoweredError, queue: -> { delayed_queue_name }
+
+  queue_as do
+    answer = arguments.first
+    question = answer.question
+
+    question.is_low_priority ? delayed_queue_name : default_queue_name
   end
 
-  def update_exp?(submission)
-    submission.assessment.autograded? && !submission.attempting? &&
-      !submission.awarded_at.nil? && submission.awarder == User.system
+  protected
+
+  def default_queue_name
+    raise NotImplementedError, 'Subclasses must implmement default_queue_name method.'
   end
 
-  def process_job_and_downgrade_if_overdue(answer)
-    process_and_measure_execution_time(answer)
+  def delayed_queue_name
+    raise NotImplementedError, 'Subclasses must implmement delayed_queue_name method.'
+  end
 
-    Course::Assessment::Submission::CalculateExpService.update_exp(answer.submission) if update_exp?(answer.submission)
+  # Performs the auto grading.
+  #
+  # @param [String|nil] redirect_to_path The path to be redirected after auto grading job was
+  #   finished.
+  # @param [Course::Assessment::Answer] answer the answer to be graded.
+  # @param [String] redirect_to_path The path to redirect when job finishes.
+  def perform_tracked(answer, redirect_to_path = nil)
+    ActsAsTenant.without_tenant do
+      raise PriorityShouldBeLoweredError if !queue_name.include?('delayed') && answer.question.is_low_priority
+
+      downgrade_if_timeout(answer.question) do
+        Course::Assessment::Answer::AutoGradingService.grade(answer)
+      end
+
+      if update_exp?(answer.submission)
+        Course::Assessment::Submission::CalculateExpService.update_exp(answer.submission)
+      end
+    end
+
+    redirect_to redirect_to_path
+  end
 
-    downgrade_if_overdue(answer.question, @execution_time)
+  def update_exp?(submission)
+    submission.assessment.autograded? && !submission.attempting? &&
+      !submission.awarded_at.nil? && submission.awarder == User.system
   end
 
-  def process_and_measure_execution_time(answer)
+  def downgrade_if_timeout(question, &block)
     start_time = Time.now
-    Course::Assessment::Answer::AutoGradingService.grade(answer)
+    block.call
+    end_time = Time.now
+    return unless !question.is_low_priority? && end_time - start_time > DEFAULT_TIMEOUT
 
-    @execution_time = Time.now - start_time
+    question.update_attribute(:is_low_priority, true)
```

Comment:
nit: to keep the code styling similar with other update attempt, I suggests using `update` instead of `update_attribute`.

Reply:
update_attribute bypasses any unnecessary validators and callbacks

Reply:
I see. In that case, the usage of `update_attribute` is correct